src/liger_kernel/chunked_loss/orpo_loss.py

```diff
@@ -0,0 +1,180 @@
+import torch
+from triton import next_power_of_2
+import torch.nn.functional as F
+from liger_kernel.ops.utils import element_mul_kernel
+
+# The hard limit of TRITON_MAX_TENSOR_NUMEL is 1048576 https://github.com/triton-lang/triton/blob/ba42a5c68fd0505f8c42f4202d53be0f8d9a5fe0/python/triton/language/core.py#L19
+# However, setting limit as 65536 as in LayerNorm tutorial is faster because of less register spilling
+# The optimal maximum block size depends on your hardware, your kernel, and your dtype
+MAX_FUSED_SIZE = 65536 // 2
+
+
+def odds_ratio_loss(chosen_logps, rejected_logps, beta=0.1):
+    """
+    Compute odds-ratio loss.
+    Args:
+        chosen_logps (torch.Tensor): Avg log probabilities of chosen tokens. Shape: (batch_size,).
+        rejected_logps (torch.Tensor): Avg log probabilities of rejected tokens. Shape: (batch_size,).
+        beta (float): Weight for the odds ratio loss.
+    """
+    log_odds = (chosen_logps - rejected_logps) - (
+        torch.log1p(-torch.exp(chosen_logps)) - torch.log1p(-torch.exp(rejected_logps))
+    )
+    ratio = F.logsigmoid(log_odds)
+    return beta * ratio.sum()
+
+
+class LigerFusedLinearORPOFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, _input, weight, target, bias=None, ignore_index=-100, beta=0.1, compiled=True):
+        """
+        Fused linear forward function with ORPO (Odds-Ratio Preference Optimization).
+        Args:
+            _input (torch.Tensor): Input tensor. Shape: (batch_size, seq_len, hidden_size).
+            weight (torch.Tensor): Weight tensor. Shape: (vocab_size, hidden_size).
+            bias (torch.Tensor, optional): Bias tensor. Shape: (hidden_size,).
+            ignore_index (int): Index to ignore for loss computation.
+            compiled (bool): Whether to use compiled mode for chunk accumulation.
+        """
+        CHUNK_SIZE = 1
+
+        def _compute_orpo_loss(input_chunk, weight, target_chunk, bias=None):
+            len_chosen_chunk = target_chunk.shape[0] // 2
+
+            unnorm_logits = input_chunk @ weight.t()  # chunk_size x V
+            if bias is not None:
+                unnorm_logits = unnorm_logits + bias
+            unnorm_logits = unnorm_logits.float()
+            norm_logits = F.log_softmax(unnorm_logits, dim=-1)
+
+            chosen_nll_loss = F.nll_loss(
+                norm_logits[:len_chosen_chunk].view(-1, norm_logits.shape[-1]),
+                target_chunk[:len_chosen_chunk].view(-1),
+                reduction="sum",
+                ignore_index=ignore_index,
+            )
+            chosen_nll_loss = chosen_nll_loss / (target[:target.shape[0]//2] != ignore_index).sum()
```

Comment:
If my understanding is correct, this is computing the mean nll loss for the chosen/accepted response, why can't we just use 'mean' as the reduction method above rather than computing the sum first and then dividing by the number of non-pad tokens?

Reply:
The target variable here is the whole target while target_chunk is the target for a given chunk. Normalizing directly using nll_loss mean reduction would normalize with target_chunk length which is mathematically wrong. So this is kinda hacky way to do normalization

Reply:
Ahh gotcha, that makes sense, thanks for the explanation.